deel/lip/normalizers.py

```diff
@@ -238,3 +240,155 @@ def spectral_normalization(
     # use sigma + eps, which ensure stability of the bjorck even when beta=0.5
     W_bar = kernel / (sigma + eps)
     return W_bar, _u, sigma
+
+
+def _power_iteration_conv(
+    w,
+    u,
+    stride=1.0,
+    conv_first=True,
+    pad_func=None,
+    eps=DEFAULT_EPS_SPECTRAL,
+    maxiter=DEFAULT_MAXITER_SPECTRAL,
+    big_constant=-1,
+):
+    """
+    Internal function that performs the power iteration algorithm for convolution.
+
+    Args:
+        w: weights matrix that we want to find eigen vector
+        u: initialization of the eigen matrix should be ||u||=1 for L2_norm
+        stride: stride parameter of the convolution
+        conv_first: RO or CO case , should be True in CO case (stride^2*C<M)
+        pad_func: function for applying padding (None is padding same)
+        eps: epsilon stopping criterion: norm(ut - ut-1) must be less than eps
+        maxiter: maximum number of iterations for the algorithm
+        big_constant: only for computing the minimum singular value (otherwise -1)
+    Returns:
+         u and v corresponding to the maximum eigenvalue
+
+    """
+
+    def identity(x):
+        return x
+
+    # If pad_func is None, standard convolution with SAME padding
+    # Else, pad_func padding function (externally defined)
+    #       + standard convolution with VALID padding.
+    if pad_func is None:
+        pad_type = "SAME"
+        _pad_func = identity
+    else:
+        pad_type = "VALID"
+        _pad_func = pad_func
+
+    def _conv(u, w, stride):
+        u_pad = _pad_func(u)
+        return tf.nn.conv2d(u_pad, w, stride, pad_type)
+
+    def _conv_transpose(u, w, output_shape, stride):
+        if pad_func is None:
+            return tf.nn.conv2d_transpose(u, w, output_shape, stride, pad_type)
+        else:
+            u_upscale = _zero_upscale2D(u, (stride, stride))
+            w_adj = _maybe_transpose_kernel(w, True)
+            return _conv(u_upscale, w_adj, stride=1)
+
+    def body(_u, _v, _old_u, _norm_u):
+        _old_u = _u
+        u = _u  # normalization should be done before
```

Comment:
seem like the normalization should be done here instead of l312

Reply:
I've changed the comment to be more understandable